### ISSUE TITLE
IRC Client Crashing

### ISSUE THREAD

**Issue by arraystock** [SerenityOS/serenity#791]
Whenever I attempt to run the IRC client, it crashes. Log:
```
year: 2019, month: 11, day: 17
PIC(i8259): cascading mode, vectors 0x50-0x58
PS2MouseDevice: Device detected
PS2MouseDevice: Mouse wheel enabled!
SB16: found version 4.5
VC: Switch to 0 (0x002c41a4)
Starting Serenity Operating System...
MM: Multiboot mmap: base_addr = 0x000000000, length = 0x00009fc00, type = 0x1
MM: Multiboot mmap: base_addr = 0x00009fc00, length = 0x000000400, type = 0x2
MM: Multiboot mmap: base_addr = 0x0000f0000, length = 0x000010000, type = 0x2
MM: Multiboot mmap: base_addr = 0x000100000, length = 0x007ee0000, type = 0x1
MM: Multiboot mmap: base_addr = 0x007fe0000, length = 0x000020000, type = 0x2
MM: Multiboot mmap: base_addr = 0x0fffc0000, length = 0x000040000, type = 0x2
MM initialized.
Initializing APIC, base: Pfee00000
Enabling local APIC for cpu #0
PIT: 1000 Hz, square wave (4a9)
PCI device: bus=0 slot=0 function=0 id=8086:1237
PCI device: bus=0 slot=1 function=0 id=8086:7000
PCI device: bus=0 slot=1 function=1 id=8086:7010
PCI device: bus=0 slot=1 function=3 id=8086:7113
PCI device: bus=0 slot=2 function=0 id=1234:1111
PCI device: bus=0 slot=3 function=0 id=8086:100e
BXVGA: framebuffer @ Pf8000000
E1000: Found at PCI address 00:03:00
E1000: IO port base: c000
E1000: MMIO base: Pfeb80000
E1000: Interrupt line: 11
E1000: Has EEPROM? 1
E1000: MAC address: 52:54:00:12:34:56
Process: New process PID=0 with name=colonel
Thread{0x00401fc8}: New thread TID=0 in colonel(0)
Process: New process PID=1 with name=init_stage2
Thread{0x004028d0}: New thread TID=0 in init_stage2(1)
Process: New process PID=2 with name=syncd
Thread{0x00403198}: New thread TID=0 in syncd(2)
Process: New process PID=3 with name=Finalizer
Thread{0x00403a58}: New thread TID=0 in Finalizer(3)
Syscall: int 0x82 handler installed
PATAChannel: PATA Controller found! id=8086:7010
PATAChannel: Bus master IDE: I/O @ c040
PATAChannel: Name="QEMU HARDDISK", C/H/Spt=247/16/63
NP(kernel): Copying new kernel mapping for V0xc09ca008 into current page directory
VFS: mounted root on Ext2FS{0x00404568}
Load ksyms
Loading ksyms...ok
Loaded ksyms
fdc: Version = 0x90
sr0 = 0xc0, cyl = 0x0
sr0 = 0xc1, cyl = 0x0
sr0 = 0xc2, cyl = 0x0
sr0 = 0xc3, cyl = 0x0
fd0 is 1.44MB floppy drive
fd1 type unsupported! Type == 0x0
Process: New process PID=4 with name=SystemServer
Thread{0x00407d20}: New thread TID=0 in SystemServer(4)
SystemServer(4) do_exec(/bin/SystemServer): thread_count() = 1
Process: New process PID=5 with name=NetworkTask
Thread{0x0040ae10}: New thread TID=0 in NetworkTask(5)
0x0013849c  Process::sys$exit(int) +14
0x001454b1  init_stage2() +2031
SystemServer(Finalizing Thread 0 in init_stage2(1)
Finalizing Process init_stage2(1)
reap: init_stage2(1) {Dead}
~Process{0x00402728} name=init_stage2 pid=1, m_fds=0
~Thread{0x004028d0}
reaped unparented process init_stage2(1), exit status: 0
4): Forking for /bin/mount...
Process: New process PID=6 with name=SystemServer
Thread{0x004028e8}: New thread TID=0 in SystemServer(6)
SystemServer(4): Process /bin/mount hopefully started with priority 3...
sys$waitpid(-1, 0x00000000, 0)
NetworkTask: LoopbackAdapter network adapter found: hw=00:00:00:00:00:00 address=127.0.0.1 netmask=255.0.0.0 gateway=0.0.0.0
NetworkTask: E1000NetworkAdapter network adapter found: hw=52:54:00:12:34:56 address=10.0.2.15 netmask=255.255.255.0 gateway=10.0.2.2
NetworkTask: Enter main loop.
SystemServer(6): Executing for /bin/mount... at prio 3
SystemServer(6) do_exec(/bin/mount): thread_count() = 1
NP(kernel): Copying new kernel mapping for V0xc0a04f2e into current page directory
mount(6): Mounting all filesystems...
mount(6): Skipping mounting root
mount(6): Mounting proc(proc) on /proc
mount proc: device proc @ /proc
VFS: Mounting ProcFS at /proc (inode: 1:67057)
mount: successfully mounted proc on /proc
mount(6): Mounting devpts(devpts) on /dev/pts
mount devpts: device devpts @ /dev/pts
VFS: Mounting DevPtsFS at /dev/pts (inode: 1:71122)
mount: successfully mounted devpts on /dev/pts
mount(6): Mounting tmp(tmp) on /tmp
mount tmp: device tmp @ /tmp
VFS: Mounting TmpFS at /tmp (inode: 1:36577)
mount: successfully mounted tmp on /tmp
0x0013849c  Process::sys$exit(int) +14
0x0013c327  syscall_trap_entry +155
0x0013c254  syscall_trap_handler +39
0x08068d21  exit +65
0x08048921  __cxa_pure_virtual +0
Finalizing Thread 0 in mount(6)
Finalizing Process mount(6)
signal 17 was ignored by SystemServer(4)
NP(kernel): Copying new kernel mapping for V0xc0a04f2e into current page directory
reap: mount(6) {Dead}
~Process{0x00402700} name=mount pid=6, m_fds=0
~Thread{0x004028e8}
SystemServer(4): Forking for /bin/TTYServer...
Process: New process PID=7 with name=SystemServer
Thread{0x004028e8}: New thread TID=0 in SystemServer(7)
SystemServer(4): Process /bin/TTYServer hopefully started with priority 3...
SystemServer(4): Forking for /bin/LookupServer...
Process: New process PID=8 with name=SystemServer
Thread{0x004123b8}: New thread TID=0 in SystemServer(8)
SystemServer(4): Process /bin/LookupServer hopefully started with priority 0...
SystemServer(4): Forking for /bin/WindowServer...
Process: New process PID=9 with name=SystemServer
Thread{0x00413da0}: New thread TID=0 in SystemServer(9)
SystemServer(4): Process /bin/WindowServer hopefully started with priority 3...
SystemServer(4): Forking for /bin/AudioServer...
Process: New process PID=10 with name=SystemServer
Thread{0x00415c08}: New thread TID=0 in SystemServer(10)
SystemServer(4): Process /bin/AudioServer hopefully started with priority 3...
SystemServer(4): Forking for /bin/Taskbar...
Process: New process PID=11 with name=SystemServer
Thread{0x00417a78}: New thread TID=0 in SystemServer(11)
SystemServer(4): Process /bin/Taskbar hopefully started with priority 3...
SystemServer(4): Forking for /bin/Terminal...
Process: New process PID=12 with name=SystemServer
Thread{0x004198a0}: New thread TID=0 in SystemServer(12)
SystemServer(4): Process /bin/Terminal hopefully started with priority 2...
SystemServer(4): Read command line: kernel hello

SystemServer(4): Continuing normally
SystemServer(8): Executing for /bin/LookupServer... at prio 0
SystemServer(8) do_exec(/bin/LookupServer): thread_count() = 1
SystemServer(9): Executing for /bin/WindowServer... at prio 3
SystemServer(9) do_exec(/bin/WindowServer): thread_count() = 1
SystemServer(10): Executing for /bin/AudioServer... at prio 3
SystemServer(7): Executing for /bin/TTYServer... at prio 3
SystemServer(7) do_exec(/bin/TTYServer): thread_count() = 1
SystemServer(12): Executing for /bin/Terminal... at prio 2
SystemServer(12) do_exec(/bin/Terminal): thread_count() = 1
SystemServer(11): Executing for /bin/Taskbar... at prio 3
Starting console server on tty1
Running shell on tty1
Process: New process PID=13 with name=TTYServer
Thread{0x0041b120}: New thread TID=0 in TTYServer(13)
sys$waitpid(13, 0x01427f8c, 0)
SystemServer(10) do_exec(/bin/AudioServer): thread_count() = 1
VFS::create: 'rpc.12' in 4:1
sys$fcntl: fd=6, cmd=3, arg=0
sys$fcntl: fd=6, cmd=4, arg=2048
Terminal(12): CLocalSocket{0x01557016} connecting to /tmp/wsportal...
Terminal(12): perror(): CSocket::common_connect: connect: Connection refused
Client::Connection: connect failed: 0, Success (not an error)
Terminal(12): CLocalSocket{0x01557016} connecting to /tmp/wsportal...
TTYServer(13) do_exec(/bin/sh): thread_count() = 1
VFS::create: 'rpc.9' in 4:1
VFS::create: 'wsportal' in 4:1
BXVGADevice resolution set to 1024x768 (pitch=4096)
BXVGA: WindowServer(9) created Region{0x002bec90} with size 6291456 for framebuffer Pf8000000 with vaddr V0x01531000
Thread{0x004288f8}: New thread TID=1 in WindowServer(9)
WindowServer(9): Started a thread, tid = 1
SystemServer(11) do_exec(/bin/Taskbar): thread_count() = 1
NP(kernel): Copying new kernel mapping for V0xc0a04eee into current page directory
/dev/tty1 set_termios: ECHO=false, ISIG=true, ICANON=false, ECHOE=false, ECHOK=false, ECHONL=false, ISTRIP=false, ICRNL=false, INLCR=false, IGNCR=false
sh -c '/bin/Shell'
</bin/Shell> 

</bin/Shell> <(null)> 
Process: New process PID=14 with name=sh
Thread{0x004100a8}: New thread TID=0 in sh(14)
Closing fds in shell process:
Now we gotta wait on children:
  21778224
sys$waitpid(14, 0x014c4e38, 6)
/dev/tty1 set_termios: ECHO=true, ISIG=true, ICANON=true, ECHOE=false, ECHOK=false, ECHONL=false, ISTRIP=false, ICRNL=false, INLCR=false, IGNCR=false
VFS::create: 'rpc.11' in 4:1
sys$fcntl: fd=6, cmd=3, arg=0
sys$fcntl: fd=6, cmd=4, arg=2048
Taskbar(11): CLocalSocket{0x01540016} connecting to /tmp/wsportal...
VFS::create: 'rpc.10' in 4:1
Thread{0x0042e1c0}: New thread TID=1 in AudioServer(10)
AudioServer(10): Started a thread, tid = 1
VFS::create: 'asportal' in 4:1
sh(14) do_exec(/bin/Shell): thread_count() = 1
/dev/tty1 set_termios: ECHO=false, ISIG=true, ICANON=false, ECHOE=false, ECHOK=false, ECHONL=false, ISTRIP=false, ICRNL=false, INLCR=false, IGNCR=false
WindowServer(9): perror(): open: Invalid argument
WindowServer(9): perror(): open: Invalid argument
Entering WindowServer main loop.
Terminal(12): CLocalSocket{0x01557016} connected ok!
Process: New process PID=15 with name=Terminal
Thread{0x00438f68}: New thread TID=0 in Terminal(15)
Taskbar(11): CLocalSocket{0x01540016} connected ok!
Terminal(15) do_exec(/bin/Shell): thread_count() = 1
/dev/pts/0 set_termios: ECHO=false, ISIG=true, ICANON=false, ECHOE=false, ECHOK=false, ECHONL=false, ISTRIP=false, ICRNL=false, INLCR=false, IGNCR=false
Terminal: Load config file from /home/anon/Terminal.ini
/dev/pts/0: Send signal 28 to everyone in pgrp 15
/dev/pts/0: Send signal 28 to Shell(15)
signal: kernel sent 28 to Shell(15)
Unblock Shell(15) due to signal
WindowServer(9): [WM] WSWindow{0x0152b806} title set to "Terminal"
WindowServer(9): WindowServer: Sharing icon buffer 1 with PID 11
LookupServer: Using network config file at /etc/LookupServer.ini.
LookupServer: Loading hosts from /etc/hosts
WindowServer(9): [WM] WSWindow{0x0152b806} title set to "anon@courage:/home/anon"
VFS::create: '.LookupServer-socket' in 4:1
<irc> 

<irc> <(null)> 
Process: New process PID=16 with name=Shell
Thread{0x00431310}: New thread TID=0 in Shell(16)
Closing fds in shell process:
Now we gotta wait on children:
  21778224
sys$waitpid(16, 0x014c4e38, 6)
/dev/pts/0 set_termios: ECHO=true, ISIG=true, ICANON=true, ECHOE=false, ECHOK=false, ECHONL=false, ISTRIP=false, ICRNL=false, INLCR=false, IGNCR=false
Shell(16) do_exec(irc): thread_count() = 1
Shell(16) do_exec(/bin/irc): thread_count() = 1
VFS::create: 'rpc.16' in 4:1
sys$fcntl: fd=6, cmd=3, arg=0
sys$fcntl: fd=6, cmd=4, arg=2048
irc(16): CLocalSocket{0x015ab016} connecting to /tmp/wsportal...
irc(16): CLocalSocket{0x015ab016} connected ok!
LookupServer: Got request for 'chat.freenode.net' (using IP 8.8.8.8)
sendto 0x014d1016 (35), flags=0, addr: 0x014c49f4 (16)
ARP table (1 entries):
52:55:0a:00:02:02 :: 10.0.2.2
sending as udp packet from 10.0.2.15:38799 to 8.8.8.8:53!
Got response (ID: 1)
  Question count: 1
  Answer count: 5
 Authority count: 0
Additional count: 0
LookupServer:     Answer #0: (question: chat.freenode.net.), type=1, ttl=119, length=4, data=185.30.166.37
LookupServer:     Answer #1: (question: chat.freenode.net.), type=1, ttl=119, length=4, data=38.229.70.22
LookupServer:     Answer #2: (question: chat.freenode.net.), type=1, ttl=119, length=4, data=94.125.182.252
LookupServer:     Answer #3: (question: chat.freenode.net.), type=1, ttl=119, length=4, data=149.56.134.238
LookupServer:     Answer #4: (question: chat.freenode.net.), type=1, ttl=119, length=4, data=64.86.243.181
irc(16): CSocket::connect: Resolved 'chat.freenode.net' to 185.30.166.37
irc(16): CTCPSocket{0x015ab20e} connecting to 185.30.166.37:0...
irc(16): CTCPSocket{0x015ab20e} connection in progress (EINPROGRESS)
WindowServer(9): [WM] WSWindow{0x01f6a806} title set to "IRC Client: anon_seren1ty@chat.freenode.net:6667"
WindowServer(9): WindowServer: Sharing icon buffer 6 with PID 11
irc(16): CTCPSocket{0x015ab20e} connected!
sendto 0x015deee3 (39), flags=0, addr: 0x00000000 (0)
sendto 0x015def23 (20), flags=0, addr: 0x00000000 (0)
IRCClient: Channels to autojoin: #serenityos
sendto 0x015deee3 (18), flags=0, addr: 0x00000000 (0)
irc(16) Couldn't find user region for V0xaaaaaab2
NP(error) fault at invalid address V0xaaaaaab2
irc(16:0) Unrecoverable page fault, read from address 0xaaaaaab2
CRASH: Page Fault Process: irc(16)
exception code: 0000
  pc=0008:0014a7e0 ds=0010 es=0010 fs=0010 gs=0010
 stk=ff4c:c0ae0e46
kstk=0010:c0ae1000, base=c0ad1000
eax=00000000 ebx=00431bd0 ecx=00000000 edx=aaaaaaaa
ebp=c0ae0de6 esp=c0ae0e46 esi=00432bd0 edi=00431bd4
irc(16) Couldn't find user region for V0x0014a7e0
0x0014a7e0  errno +1353692
0x001385f5  Process::crash(int, unsigned int) +183
0x001110ee  exception_6_handler +0
0x00111257  exception_14_handler +289
0x0011068b  exception_14_entry +37
0x0012ff4c  TCPSocket::send_outgoing_packets() +244
0x0013013c  TCPSocket::send_tcp_packet(unsigned short, void const*, int) +462
0x001301d6  TCPSocket::protocol_send(void const*, int) +22
0x0012a8ff  IPv4Socket::sendto(FileDescription&, void const*, unsigned long, int, sockaddr const*, unsigned int) +241
0x00133511  Process::sys$sendto(Syscall::SC_sendto_params const*) +203
0x0013c45a  syscall_trap_entry +462
0x0013c254  syscall_trap_handler +39
0x080ba00a  send +24
0x08095bc1  _ZN7CSocket4sendERKN2AK10ByteBufferE +37
0x08048eb8  _ZN9IRCClient4sendERKN2AK6StringE +106
0x080493eb  _ZN9IRCClient12join_channelERKN2AK6StringE +47
0x080498f8  _ZN9IRCClient19on_socket_connectedEv +410
0x08095e1d  _ZNK2AK8FunctionIFvvEE15CallableWrapperIZN7CSocket14common_connectEPK8sockaddrjEUlvE_E4callEv +119
0x0809858a  _ZN7CObject14dispatch_eventER6CEventPS_ +86
0x0809a115  _ZN10CEventLoop4pumpENS_8WaitModeE +477
0x0809a2ad  _ZN10CEventLoop4execEv +43
0x0806a03f  _ZN12GApplication4execEv +29
0x08048389  main +81
0x08048a89  _start +94
Process irc(16) regions:
BEGIN       END         SIZE        ACCESS  NAME
01001000 -- 0118cfff    0018c000    R       /bin/IRCClient
08048000 -- 0810dfff    000c6000    R X     elf-map-rx
0810e000 -- 0810ffff    00002000    RW      elf-alloc-rw
0118e000 -- 0118efff    00001000    RW      (null)
01190000 -- 01190fff    00001000    RW      Thread-specific
01192000 -- 01591fff    00400000    RW      Stack (Main thread)
01593000 -- 01593fff    00001000    RW      malloc: ChunkedBlock(8)
01595000 -- 01595fff    00001000    RW      malloc: ChunkedBlock(32)
01597000 -- 01597fff    00001000    RW      malloc: ChunkedBlock(16)
01599000 -- 01599fff    00001000    RW      malloc: ChunkedBlock(64)
0159b000 -- 0159bfff    00001000    RW      malloc: ChunkedBlock(128)
0159d000 -- 0159dfff    00001000    RW      malloc: BigAllocationBlock
0159f000 -- 0159ffff    00001000    RW      malloc: ChunkedBlock(1016)
015a1000 -- 015a1fff    00001000    RW      malloc: ChunkedBlock(16)
015a3000 -- 015a3fff    00001000    RW      malloc: ChunkedBlock(32)
015a5000 -- 015a5fff    00001000    RW      malloc: ChunkedBlock(128)
015a7000 -- 015a7fff    00001000    RW      malloc: ChunkedBlock(64)
015a9000 -- 015a9fff    00001000    RW      malloc: ChunkedBlock(8)
015ab000 -- 015abfff    00001000    RW      malloc: ChunkedBlock(252)
015ad000 -- 015adfff    00001000    RW      malloc: BigAllocationBlock
015af000 -- 015affff    00001000    RW      malloc: ChunkedBlock(508)
015b1000 -- 015b1fff    00001000    RW      malloc: ChunkedBlock(2036)
015b3000 -- 015b3fff    00001000    R       SharedBuffer
015b9000 -- 015bbfff    00003000    R       /res/fonts/Katica10.font
015cf000 -- 015cffff    00001000    R       SharedBuffer
015d1000 -- 015d1fff    00001000    R       SharedBuffer
015bd000 -- 015bdfff    00001000    R       SharedBuffer
015d3000 -- 015d3fff    00001000    R       SharedBuffer
015c4000 -- 015c4fff    00001000    R       SharedBuffer
015c6000 -- 015c6fff    00001000    R       SharedBuffer
015cc000 -- 015ccfff    00001000    RW      malloc: ChunkedBlock(252)
015d5000 -- 015d7fff    00003000    R       /res/fonts/KaticaBold10.font
015d9000 -- 015d9fff    00001000    RW      malloc: ChunkedBlock(32)
015de000 -- 015defff    00001000    RW      malloc: ChunkedBlock(64)
015e0000 -- 015e2fff    00003000    R       /res/fonts/CsillaThin7x10.font
015e6000 -- 015e6fff    00001000    RW      GraphicsBitmap [16x16] - Decoded PNG: /res/icons/16x16/undo.png
015e8000 -- 015e8fff    00001000    RW      GraphicsBitmap [16x16] - Decoded PNG: /res/icons/16x16/redo.png
015ec000 -- 015ecfff    00001000    RW      GraphicsBitmap [16x16] - Decoded PNG: /res/icons/cut16.png
015ee000 -- 015eefff    00001000    RW      GraphicsBitmap [16x16] - Decoded PNG: /res/icons/16x16/edit-copy.png
015f2000 -- 015f2fff    00001000    RW      GraphicsBitmap [16x16] - Decoded PNG: /res/icons/paste16.png
015f4000 -- 015f4fff    00001000    RW      GraphicsBitmap [16x16] - Decoded PNG: /res/icons/16x16/delete.png
015db000 -- 015dbfff    00001000    RW      malloc: ChunkedBlock(32)
015f6000 -- 015f6fff    00001000    RW      malloc: ChunkedBlock(252)
015f8000 -- 016e2fff    000eb000    RW      SharedBuffer
016e4000 -- 017cefff    000eb000    RW      SharedBuffer
Finalizing Thread 0 in irc(16)
Finalizing Process irc(16)
signal 17 was ignored by Shell(15)
reap: irc(16) {Dead}
~Process{0x0040f870} name=irc pid=16, m_fds=0
~Thread{0x00431310}
sh: command finished in 1335 ms
/dev/pts/0 set_termios: ECHO=false, ISIG=true, ICANON=false, ECHOE=false, ECHOK=false, ECHONL=false, ISTRIP=false, ICRNL=false, INLCR=false, IGNCR=false
Connection: Client disconnected: 3
WindowServer(9): Event type 3012 with no receiver :(
```

**Comment by awesomekling** [comment]
Interesting, looks like a use-after-free in the TCP sockets code (`kfree()` fills freed memory with `0xaa`).

cc @deoxxa

**Comment by awesomekling** [comment]
Oh, oh, this was fixed by 61f611bf3cbdaa726a86c14524912ebff591d868
